src/cli/rustup_mode.rs

```diff
@@ -1159,9 +1166,75 @@ fn component_add(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
         let new_component = Component::new_with_target(component, false)
             .unwrap_or_else(|| Component::new(component.to_string(), target.clone(), true));
 
-        toolchain.add_component(new_component)?;
-    }
+        if m.is_present("allow-downgrade") {
+            let toolchain_date = toolchain
+                .get_manifest()
+                .ok()
+                .and_then(|m| m.map(|m| m.date));
```

Comment:
Why not unwrap here?
```suggestion
                .and_then(|m| m.map(|m| m.date))
                .unwrap();
```